### ISSUE TITLE
Current API for defining fixtures doesn't allow fixtures to be typechecked

### ISSUE THREAD

**Issue by andreypopp** [react-cosmos/react-cosmos#560]
I think the current API is not optimal for codebases which use either Flow or TS because it doesn't allow to check fixtures for type errors.

Instead I'm proposing to have a `createFixture` function:
```
function createFixture<P: {}, C: React.ComponentType<P>>(fixture: {
  component: C,
  name?: string,
  props?: P,
  children?: React.Node,
}) {
  return fixture;
}
```

which provide a type witness to Flow/TS how `props` relates to `component`.

Which would make fixtures look like:
```
import {createFixture} from 'react-cosmos'
import MyComponent from '../MyComponent'

export default createFixture({
  component: MyComponent,
  props: {
    label: 'Hello, World'
  }
})
```

Digging Cosmos more and more! Thanks!

**Comment by skidding** [comment]
> which provide a type witness to Flow/TS how props relates to component.

Is that a working example? 

I've been wondering if there's a way to have a generic `Fixture` type that can be used like this: `fixture: Fixture<Component>` and it would know to expect `fixture.props` to match the Props type of Component. But I don't have enough experience with Flow/TS to understand if that's possible.

Btw, I just typed some fixtures for the first time today :P

```js
// @flow
import MyComponent from '../../MyComponent';
import type { Props } from '../../MyComponent';

const fixture: { props: Props } = {
  component: MyComponent,
  props: {
    // secret sauce...
  }
};

export default fixture;
```

Not ideal. Happy to continue discussion on making this more straightforward.

**Comment by andreypopp** [comment]
> Is that a working example?

Yes. I've copy pasted this from my fixture code.

> I've been wondering if there's a way to have a generic Fixture type that can be used like this: fixture: Fixture<Component> and it would know to expect fixture.props to match the Props type of Component. But I don't have enough experience with Flow/TS to understand if that's possible.

The problem with this proposed API is that the type annotation with `Fixture` is optional while using `createFixture` guarantees that fixtures will be type checked.

**Comment by maxsalven** [comment]
@andreypopp I tried using the function you defined, and it works great for type checking (although VSCode shows the errors in the component definition rather than in the fixture, but not a huge deal). However, any fixture I have wrapped in `createFixture` no longer shows up in the correct place in the file tree on the left side of Cosmos. Have you had this issue?

@skidding, if I run `yarn print-fixture-files` then it simply shows `"components": []` for the impacted components.

**Comment by skidding** [comment]
Closing this as it's now possible to use @andreypopp's `createFixture` function to typecheck fixture without experiencing the tree view issues that @maxsalven is mentioning above. That was fixed in #591. I've been actually using `createFixture` succesfully in my projects, but it's not part of the Cosmos APIs at the moment.
